libraries/chain_kv/include/b1/session/session.hpp

```diff
@@ -0,0 +1,1244 @@
+
+   template <typename Persistent_data_store_, typename Cache_data_store_>
+   friend session<Persistent_data_store_, Cache_data_store_> make_session();
+
+   template <typename Persistent_data_store_, typename Cache_data_store_>
+   friend session<Persistent_data_store_, Cache_data_store_> make_session(Persistent_data_store_ store);
+
+   template <typename Persistent_data_store_, typename Cache_data_store_>
+   friend session<Persistent_data_store_, Cache_data_store_> make_session(Persistent_data_store_ store,
+                                                                          Cache_data_store_      cache);
+
+   template <typename Persistent_data_store_, typename Cache_data_store_>
+   friend session<Persistent_data_store_, Cache_data_store_>
+   make_session(session<Persistent_data_store_, Cache_data_store_>& the_session);
+
+   // Defines a key ordered, cyclical iterator for traversing a session (which includes parents and children of each
+   // session). Basically we are iterating over a list of sessions and each session has its own cache of key_values,
+   // along with iterating over a persistent data store all the while maintaining key order with the iterator.
+   template <typename Iterator_traits>
+   class session_iterator final {
+    public:
+      using difference_type   = typename Iterator_traits::difference_type;
+      using value_type        = typename Iterator_traits::value_type;
+      using pointer           = typename Iterator_traits::pointer;
+      using reference         = typename Iterator_traits::reference;
+      using iterator_category = typename Iterator_traits::iterator_category;
+      friend session;
+
+    public:
+      session_iterator()                           = default;
+      session_iterator(const session_iterator& it) = default;
+      session_iterator(session_iterator&&)         = default;
+
+      session_iterator& operator=(const session_iterator& it) = default;
+      session_iterator& operator=(session_iterator&&) = default;
+
+      session_iterator& operator++();
+      session_iterator  operator++(int);
+      session_iterator& operator--();
+      session_iterator  operator--(int);
+      value_type        operator*() const;
+      value_type        operator->() const;
+      bool              operator==(const session_iterator& other) const;
+      bool              operator!=(const session_iterator& other) const;
+
+    protected:
+      template <typename Test_predicate, typename DB_test_predicate, typename Move_predicate>
+      void move_(const Test_predicate& test, const DB_test_predicate& db_test, const Move_predicate& move);
+      void move_next_();
+      void move_previous_();
+
+    private:
+      typename iterator_cache_type::iterator m_active_iterator;
+      session*                               m_active_session{ nullptr };
+   };
+
+   struct iterator_traits final {
+      using difference_type     = long;
```

Comment:
Why is the difference_type a `long` this is likely to be 32-bits was that intended?  I don't fully understand what the difference of a `session_iterator` would mean.

Reply:
Copy and paste errors.  Updated.